Cargo.toml

```diff
@@ -93,8 +93,9 @@ wgpu = { version = "22.0.0", default-features = false, features = [
     # Make the renderer `Sync` even on wasm32, because it makes the code simpler:
     "fragile-send-sync-non-atomic-wasm",
 ] }
-winit = { version = "0.29.4", default-features = false }
 
+# Version 0.30.1 and up unfortunately crash on macOS, due to https://github.com/rust-windowing/winit/pull/3684
+winit = { version = "=0.30.0", default-features = false }
```

Comment:
Unfortunately, by pinning to 0.30.0 we have the problem that https://github.com/rust-windowing/winit/pull/3684 solves, which is crashes during shutdown.

Running this branch gives me this one exit:
![image](https://github.com/user-attachments/assets/43bb6313-380a-4297-80e1-7ccddb931510)

Reply:
Ah crap :/ That weirdly doesn't repro for me - either by closing windows or ctrl-c'ing but well, it's definitely missing the bugfix so makes sense....

I tried some more workarounds for 0.30.4 but really can't find anything that prevents the crash. I guess that leaves a few options:

- Go with 0.30.0 and crash on exit on mac
- Go with 0.30.4 and crash on immediate viewports (perhaps with an error)
- Rejig immediate viewports to be more like deferred somehow
- Wait for a patched winit release / workaround to be available

Lmk your thoughts!

Reply:
>Go with 0.30.0 and crash on exit on mac

Having eframe always crash on exit on Mac is not an option imho.

> Go with 0.30.4 and crash on immediate viewports (perhaps with an error)

This is better, but not great. Users _can_ switch to using deferred viewports if they work. We could guide users towards it gently (with a log message) or harshly (with `#[cfg(not(target_os = "macos"))]` on `show_viewport_immediate`)

> Rejig immediate viewports to be more like deferred somehow

If we can circumvent the bug somehow that would be great, but seems unlikely?

> Wait for a patched winit release / workaround to be available

Waiting won't give us much unless someone volunteers to work on https://github.com/rust-windowing/winit/issues/3800

Reply:
Thank you for that! All good points.

Good news though :) I gave it one last go as you're right someone has to volunteer, and noticed this time that the crash complains about a reference which just so happens to come from an unsafe block not too far up in the call stack from egui....

Integrations register a callback for immediate viewports. When registering this gets an &ActiveEventLoop, then uses unsafe to cast away the lifetime, and store that pointer in the callbak. That's really not allowed! At any point there might be an aliased mutable reference, and the comment about how the lifetime outlives the callback doesnt really make sense to me - winit is free to do what it wants! The value could even just be moved in the meantime? This probably worked by happenstance on older winit versions.

So, instead I now make a tiny thread_local context which stores the pointer for the duration of the painting. I'm not very happy with it, this still leaves an unsafe block, but still seems much easier to reason about. See the safety comment for some justification, though I've never PR'ed unsafe code before so beware :) 

To really get rid of unsafe require a bigger rework that would get rid of this depencency injection shenannigans entirely I think.

Reply:
Nicely done!